### PR TITLE
Make project work and add Dockerfile

### PR DESCRIPTION
**Pull request by NullIsNot0**
This PR fixes error where application was not found and adds Dockerfile to build and run solution for easier project adoption for newcommers.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+FROM alpine:3.16 AS production
+
+ARG CORERULESET_VERSION=v4.0.0-rc1
+ARG CORERULESET_MD5=9140236dc7e941c274e414385824c996
+
+COPY ./docker/coraza-spoa/coraza.conf /etc/coraza-spoa/coraza.conf
+COPY ./docker/start.sh ./
+COPY ./config.yaml.default ./config.yaml
+
+RUN set -eux; \
```
Comment (anuraaga) [inline Dockerfile:28]: Can this also happen in the builder?

File: docker-compose.yaml [hunk 2]
```diff
@@ -2,10 +2,10 @@ version: "3.9"
 services:
   coraza:
     build: .
-    volumes:
-      - type: bind
-        source: ./docker/coraza-spoa
-        target: /etc/coraza-spoa
+    # volumes:
```
Comment (anuraaga) [inline docker-compose.yaml:5]: If these aren't needed can you remove the lines?

File: internal/response.go [hunk 3]
```diff
@@ -59,7 +59,7 @@ func (s *SPOA) processResponse(msg spoe.Message) ([]spoe.Action, error) {
 				break
 			}
 
-			if tx, ok = txInterface.(*coraza.Transaction); ok {
+			if tx, ok = txInterface.(*coraza.Transaction); !ok {
```
Comment (anuraaga) [inline internal/response.go:62]: 🙀

Comment (NullIsNot0) [reply internal/response.go:62]: Is this 🙀 good or bad?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,49 @@
+FROM golang:1.19.1-alpine3.16 AS builder
+
+WORKDIR /app
+RUN set -eux; \
```
Comment (jcchavezs) [inline Dockerfile:4]: Do you mind adding a comment in the code explaining why all these deps should be installed?

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    apk add --no-cache \
+        gc-dev \
+        make \
+        gcc \
+        git \
+        musl-dev \
+        pcre-dev \
+    ;
+
+COPY . .
```
Comment (jcchavezs) [inline Dockerfile:14]: It is a good idea to first copy the go.mod and go.sum, run go get ./... and then copy the rest of the code. This way you will cache the pull of deps.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+RUN make
+
+
+
+FROM alpine:3.16 AS production
```
Comment (jcchavezs) [inline Dockerfile:19]: nit: the `production` name isn't too friendly to the 12 factor mindset. How about not giving any name to it?

File: docker/start.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env sh
```
Comment (jcchavezs) [inline docker/start.sh:1]: Was this script copied from somewhere? If so, could we link the original? if not, could we add comments on what exactly we are doing on every line as it looks very arbitrary and hard to digest?

File: internal/response.go (hunk 3 above)
Comment (jcchavezs) [reply internal/response.go:62]: 🙀

Comment (anuraaga) [reply internal/response.go:62]: Sorry for the confusion - it's good 👍

File: Dockerfile [hunk 8]
```diff
@@ -14,41 +17,43 @@ RUN set -eux; \
...
-    && cp /tmp/coraza-coreruleset/rules/* /etc/coraza-spoa/rules/ \
-    && rm -rf /tmp/coraza-coreruleset \
-    && find /etc/coraza-spoa/rules -type f -name '*.example' | while read -r f; do cp -p "$f" "${f%.example}"; done \
-    && sed -i.example 's/^\(SecDefaultAction "phase:[12]\),log,auditlog,pass"/\1,log,noauditlog,deny,status:403"/' /etc/coraza-spoa/crs-setup.conf
+    && mv /tmp/coraza-coreruleset/crs-setup.conf.example /etc/coraza-spoa/crs-setup.conf \
+    && mv /tmp/coraza-coreruleset/rules /etc/coraza-spoa \
+    && rm -rf /tmp/*
 
+# Copy project files into build image
 COPY . .
```
Comment (jcchavezs) [inline Dockerfile:31]: Please do check https://github.com/corazawaf/coraza-spoa/pull/34#discussion_r970558014

**Comment by jcchavezs** [comment]
Please do include this build in the CI to verify that further changes don't break it.

File: Dockerfile (hunk 8 above)
Comment (NullIsNot0) [reply Dockerfile:29]: I have fixed this in the latest Dockerfile commit.

**Comment by NullIsNot0** [comment]
I have no idea how to include this build in CI. Can I get any help?

**Comment by sts** [comment]
Shall Coreruleset really be included in the Docker Image? IMO it could add a lot of complexity; I did the same initially, but in the end you would have to override a lot of files I guess. Also versioning images gets much more extensive. I would actually like to drop it.

This was why I didn't release my image yet, which already includes a multi-arch build and github workflow to push images to ghcr.io! https://github.com/sts/coraza-spoa/tree/docker_build2

**Comment by NullIsNot0** [comment]
With Coreruleset inclusion my intension was to make easier project adoption for newcommers. People clone the repo, build it and can start testing it right away. If we want, we can rename `Dockerfile` to `Dockerfile.example` or `Dockerfile.with-crs` and continue to maintain it, but for those who prefer to bring their own CRS we can create `Dockerfile` and document the process of bind mounting their CRS rules.
I plan to run `coraza-spoa` containers in Kubernetes bind mounting `config.yaml` file with all the applications configured in it. Individual applications can looad their own rule files and override files (also bind mounted). And Kubernetes can horizontally scale and load balance `coraza-spoa` deployment on-demand. I haven't set it up yet, but I think I will prefer CRS built into container image and bind mount only override files. Of course, if we have official `coraza-spoa` image without CRS, I can take it as a base image and fetch CRS into it. I'm not the project maintainer to make a decision like this, I can only show my point of view.

**Comment by jptosso** [comment]
LGTM! thank you for your contribution

**Pull request merged**